app/cdap/components/PipelineDiff/util/helpers.ts

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import { objectQuery } from 'services/helpers';
+import { IPipelineStage } from '../types';
+
+export function generatePluginMapKey(node) {
+  const plugin = node.plugin;
+  const type = node.type || plugin.type;
+
+  return `${plugin.name}-${type}-${plugin.artifact.name}-${plugin.artifact.version}-${plugin.artifact.scope}`;
+}
+
+export function getCustomIconSrc(pluginsMap, pluginKey) {
+  const iconSourceType = objectQuery(pluginsMap, pluginKey, 'widgets', 'icon', 'type');
+
+  if (iconSourceType === 'inline') {
+    return objectQuery(pluginsMap, pluginKey, 'widgets', 'icon', 'arguments', 'data');
+  }
+
+  return objectQuery(pluginsMap, pluginKey, 'widgets', 'icon', 'arguments', 'url');
+}
+
+export function shouldShowAlertsPort(pluginsMap, pluginKey) {
+  return objectQuery(pluginsMap, pluginKey, 'widgets', 'emit-alerts');
+}
+
+export function shouldShowErrorsPort(pluginsMap, pluginKey) {
+  return objectQuery(pluginsMap, pluginKey, 'widgets', 'emit-errors');
+}
+
+export const getKeyFromPluginProps = (pluginProperties) => {
+  const key = pluginProperties['0'];
+  return key ? key.split('.')[1] : '';
+};
+export function getArtifactKey(artifact: IPipelineStage['plugin']['artifact']) {
+  return `${artifact.name}-${artifact.version}-${artifact.scope}`;
+}
+
+export function createPluginInfo(plugin: IPipelineStage['plugin']) {
+  const pluginKey = `${plugin.name}-${plugin.type}`;
+  const availablePluginKey = `${pluginKey}-${getArtifactKey(plugin.artifact)}`;
+
+  const info = Object.assign({}, plugin.artifact, {
+    properties: [`widgets.${pluginKey}`, `doc.${pluginKey}`],
+  });
+
+  return {
+    info,
+    key: availablePluginKey,
+  };
+}
+export const pluginDiffColors = {
+  '+': '#0D8147',
+  '-': '#BB0000',
+  '~': '#BB8700',
+  default: '#000000',
```

Comment:
something about defining the diff indicator 'default' is odd to me

Reply:
updated